rclpy/src/rclpy/_rclpy.c

```diff
@@ -587,6 +590,15 @@ rclpy_init(PyObject * Py_UNUSED(self), PyObject * args)
     rcl_init_options_t init_options = rcl_get_zero_initialized_init_options();
     rcl_ret_t ret = rcl_init_options_init(&init_options, allocator);
     if (RCL_RET_OK == ret) {
+      // Set domain id
+      ret = rcl_init_options_set_domain_id(&init_options, (size_t)domain_id);
```

Comment:
Adding a check for `domain_id > 0` here too doesn't sound like a bad idea.
I always prefer doing the check in the closest place to the unsafe cast.

Reply:
@ivanpauno 

Agree. But there is a limitation here.   
Please look at below codes

In context.py,  
```python
class Context:
  ...
  def init(...): 
     ...
        if domain_id is None:
            domain_id = rclpy_implementation.rclpy_get_default_domain_id() # return SIZE_MAX
        else:
            if domain_id < 0:
                raise RuntimeError(
                    'Domain id ({}) should never be lower than zero.'
                    .format(domain_id))
       ...
            rclpy_implementation.rclpy_init(
                args if args is not None else sys.argv,
                capsule,
                domain_id)
```

If user doesn't specify domain id, we will get default domain id which is defined as `SIZE_MAX`.   
In _rclpy.c, use `PyArg_ParseTuple()` to get parameter from python. For domain_id, if use 'L' (convert C long long), while domain_id is default value,  rclpy_implementation.rclpy_init() report error "OverflowError: int too big to convert".   
So we have to think domain_id  passing to `rclpy_init()` is unsigned long long.

Reply:
@Barry-Xu-2018 argh, good point.

I would prefer if `SIZE_MAX` in the C side is converted to `None` in the `Python` side.
`None` expresses intent better in this case.

- `rclpy_get_default_domain_id` won't be needed.
- `rclpy_init` will have to parse an optional long long ("OO|K" is the required format string). `domain_id` should be initialized to `SIZE_MAX` (`PyArg_ParseTuple` doesn't modify optional arguments when no specified).

@Barry-Xu-2018 does that make sense to you?

Reply:
@ivanpauno 

Thanks for sharing your thoughts.   
Agree. I will update codes.